### ISSUE TITLE
Make Curriculum Markdown Content Editable without knowledge of Github

### ISSUE THREAD

**Issue by emmairwin** [mozilla/participation-org#230]
Volunteer subtask of #216  

Add an 'Improve this content' button to all curriculum pages on the Education Portal.
Github:  https://github.com/emmairwin/mozilla-community-edu
# Description

If we can make this work, then it's a huge opportunity to bring more people into contribute curriculum, who feel less comfortable with Github.   We have people who want to help - we need to let them.

I really, really want to make this work 'in site' vrs redirect to something like Prose, so want to explore the in-page wiki-edit type of solution first.
### Brief description of task.
- Add a 'Improve this content' button on all curriculum pages, that allow in-site editing of content, that in turn submits a pull request.
### User Workflow

 *User views curriculum page, at top and bottom are a button 'improve this' or 'make suggestion' (not sure which sounds better)
- User must be logged in to Github.
- When  user clicks on that button, content opens in a simple markdown editor.
- If a user makes changes and clicks 'save' a pull request is submitted for those changes.
## Prerequisite

No specific per-requisite except a curiosity in how to make this work, and have a basic understanding of Github pages.
## Participation Steps
1. Read this Issue:  https://github.com/phillipadsmith/getting-down-with-github/issues/14
2. Tour these examples:  https://talk.jekyllrb.com/t/an-idea-for-front-end-editing-in-jekyll-via-the-github-api-and-web-based-markdown-editors/737/6
3.  Use this '[markdown_edit'](https://github.com/emmairwin/mozilla-community-e
   ![suggest]%28https://cloud.githubusercontent.com/assets/60618/11190710/fb572b68-8c4b-11e5-95b6-83cc8c94c163.png%29
   du/branches) feature branch to begin development, we can also collaborate here for discussion.
4. Leave your name here to let us know you are working on this, you can also ask me questions on the solution in #2, which I hacked on for a bit at Mozfest, and didn't quite get working (see screenshot of button on curriculum)

Please use comments to ask questions if you / I'm sure you'll have a few questions!

**Comment by asdofindia** [comment]
I think we have come a whole circle and are now trying to create a wiki software by hacking together jekyll. Are we actually looking for a markdown based wiki software? If that's the case there is a [markdown extension](https://www.mediawiki.org/wiki/Extension:MarkdownSyntax) for mediawiki.

One difference between that and this I see is that if we build it on top of jekyll+github pages, each change will be a pull request. (There are [extensions that allow this kind of approval-before-edit-goes-live approach](https://www.mediawiki.org/wiki/Extension:Approved_Revs) in mediawiki too, but let's set that aside).

But I wouldn't call that a bad idea. If coupled with [having a github organization for community education](https://discourse.mozilla-community.org/t/create-a-github-organization-for-community-education/5109), this will make great sense.

I think there's a slight misalignment between these two requirements

> It's also a great entry point to Github, in recognizing the power of a pull request. 

and

> make this work 'in site'

If we can drop the latter requirement, it will become easier technologically (we can just link people to the edit feature of github itself) and people will actually be exploring github and learning what they're doing. In fact, I don't get the whole point behind the exercise of replicating the exact same functionality that github web provides on our own website.

**Comment by emmairwin** [comment]
@asdofindia I would like to say that it does not need to be Jekyll, but it needs to be able to easily pull in external markdown content, and allow edit-suggest style functionality from a user-friendly UI.
- *sort of \* agree on misalignment, I would say the learning side of things is an after thought rather than a focus, so will just remove that. 
- Greg has also suggested Federated Wiki, which apparently allows editing -> pull request workflow. and supports markdown.  

Really this is a research task, I will update that topic as well.

**Comment by emmairwin** [comment]
@asdofindia 
RE: 
_One difference between that and this I see is that if we build it on top of jekyll+github pages, each change will be a pull request. (There are extensions that allow this kind of approval-before-edit-goes-live approach in mediawiki too, but let's set that aside)._

That's OK.

RE: 
_In fact, I don't get the whole point behind the exercise of replicating the exact same functionality that github web provides on our own website._

Some background.
-  I have people reaching out who are experts in instructional design, who want to edit and write curriculum but can't - because Github functionality is to hard, and not a skill they even want to learn at this moment.  Maybe eventually, but not to fix a typo.
- If I want people to edit pages in Github I have to make them a collaborator, but even then they feel nervous editing in the UI, don't understand revision history, nor should they have to.
- The repo is hard to navigate, if you don't know exactly where to look for things (this is something else I need to work on)
- Educational content in wikis is ugly (sorry), does not engage and feels like documentation. I want to fight that tendency to treat curriculum like text, when it should feel like an experience.  (which is why  currently using Course in a Box.  not married to the platform.
- Jekyll allows us to use a git submodule for curriculum, which abstracts code from content.
-  I wouldn't say we are replicating it, but obfuscating it for reasons that people simply don't need to understand some of the functionality in Github to be successful in the thing they want to do (and what we need them to do).  It's no different than using an API really.

Good challenges, I hope my answers are making sense.

**Comment by asdofindia** [comment]
Hey, I've sent a PR on that branch to demonstrate what [github's own editing](https://help.github.com/articles/editing-files-in-your-repository/) can do. See how helpful that is. (It works correctly only inside the reps_mentor_training module). 

> If I want people to edit pages in Github I have to make them a collaborator, but even then they feel nervous editing in the UI, don't understand revision history, nor should they have to. 

[People without write access will be presented with the pull request workflow](https://help.github.com/articles/editing-files-in-another-user-s-repository)

> The repo is hard to navigate, if you don't know exactly where to look for things (this is something else I need to work on)

Yes, this is correct. Even I had difficulty in finding out where things were just now. But this is just a matter of some housekeeping and following best standards (I'll help you with this). And take it from me for granted that any kind of implementation of this "magic edit button" _will require_ necessarily that **all the modules are organized in a predictable/convention-based URL pattern corresponding to the curriculum's URL** and not the way they're organized now (with includes). 

> Educational content in wikis is ugly (sorry), does not engage and feels like documentation. I want to fight that tendency to treat curriculum like text, when it should feel like an experience. (which is why currently using Course in a Box. not married to the platform.
> Jekyll allows us to use a git submodule for curriculum, which abstracts code from content.

+1. These are the plus points of using github based curriculum management. 

> but obfuscating it for reasons that people simply don't need to understand some of the functionality in Github to be successful in the thing they want to do

Check out the button I added in the PR and please see if that solves this issue.

**Comment by jgmac1106** [comment]
What if we need to rethink the simplifying of GitHub and recognize it as the next step in participation on-boarding?

So there would be tiers of curriculum development.

Tier One: Any markdown file that started with the template. You seed it everywhere. A Google Doc template, an etherpad template, a github pages template, a thimble project, the bathroom at Denny's. Everywhere. At this tier you want the widest net possible. 

Tier Two: Community members curate tier one level stuff. If something looks good they contact the author (increasing participation) and ask if they want to remix it together.

Tier Three: If tier two curriculum meets specific guidelines (accessibility, images, etc) then it can be elevated to canon. At that point the original author is given some 1:1 or small group coaching on how to use Git.

**Comment by asdofindia** [comment]
Emma, you said in emmairwin/mozilla-community-edu#38 that

> when I used my test account, I got the message that I needed to fork the repo and propose changes from there. This will scare some off our target contributor

This is a one click process that's one time per user. (After a user forks the repository once, they are not presented with "Fork this?" scary button again for subsequent edits). **Compare** this with other solutions below.

I understand all the UX gain you're wishing for as in https://talk.jekyllrb.com/t/an-idea-for-front-end-editing-in-jekyll-via-the-github-api-and-web-based-markdown-editors/737

But, with that, although the editing becomes _in-place_ and extremely intuitive UI-wise, the authentication with github might be an equally worthy downside.

Here're the options for authentication in in-place editing:
- [Make the user generate](https://help.github.com/articles/creating-an-access-token-for-command-line-use/#creating-a-token) an oauth token and save it on our site as in [wiki-hub/jekyll-collab](https://github.com/wiki-hub/jekyll-collab)  (I hope this is going to be _very_ scary for people who find "fork this repository" scary)
- Have a server that requests [oauth token using the github api](https://developer.github.com/v3/oauth/#web-application-flow). This one can be seen in use on any site that has "log in with github" option. Here, the flow would be roughly, for first run:
  - User clicks suggest changes
  - Github login/sign-up
  - An "authorize this application to access data" page. Demo: [screenshots & docs](https://help.github.com/articles/connecting-with-third-party-applications/), [in-action](https://developer.mozilla.org/users/github/login/?next=%2Fen-US%2F) (This step should be **compare**d with "fork this repository" step)
  - If user completes that step, our server is called back with oauth token which we then can save and use for next time onwards.
- Anonymous editing as in [wiki-hub/jekyll-anon](https://github.com/wiki-hub/jekyll-anon) which is pretty much the easiest workflow, but I'm not sure [how much we've to dumb this down](https://github.com/wiki-hub/jekyll-anon/pull/39).

I checked out federated-wiki too. It seems more like an experiment in federation of server than about wiki. 
[The way the changes are stored](https://github.com/reclaimhosting/federated-wiki/tree/master/twfy.stancer.federatedwiki.org/pages) it doesn't look like the purpose of a repository storing curriculum content separately is possible with federated wiki.

If I understand you correctly @jgmac1106, you suggested it to show how the front-end based markdown editing works. But, this is cheating :P This totally circumvents github, pull requests, commits, everything. Git is used, only by the server at the very back end. We may as well use mediawiki+markdown if we're going down this approach :D

## 

> What if we need to rethink the simplifying of GitHub and recognize it as the next step in participation on-boarding?

I think this is the right question to ask. What are we actually trying to do here? Do we think that people will be able to write an entire curriculum from within the front-end. Won't they want to save things in between? How will we cater to that? (Making them learn git+github and local cloning, etc definitely sounds like a worthy investment for someone who wants to _meaningfully_ contribute to curriculum). If we are making all this UX-candy just for someone to fix a typo easily, we're wasting a lot of time.

Rather, a multi-tiered, more engaging approach focussing on long-term contribution like @jgmac1106 suggested makes a lot more sense to me. 

I'd written this some time before too. Curriculum can be developed even on the back of an envelope or an etherpad. **While we're currently separating content from the portal by git submodule, are we trying to tie down content creation to the portal?**

**Comment by emmairwin** [comment]
akshay please trust me that what feels like a small one time thing for us -
is a absolute  show stopper for others.  Just the terminology can stop
people never mind the action m  I want to dream big here for making
contribution easier.

**Comment by asdofindia** [comment]
Okay, so which kind of authentication do you think we should be using? (oauth web flow, or anonymous?)

Edit: or both?

**Comment by asdofindia** [comment]
BTW, if and when we are going down in-place editing route, we don't have to reinvent the wheel. We can re-use the [github API wrapper code](https://github.com/michael/github) that powers Prose itself to do this.

**Comment by emmairwin** [comment]
Exactly and perfect!  We'll get there. Thanks for sticking with us :)

**Comment by erlend-sh** [comment]
Feel free to ping @mushishi78 directly about the prospects using wiki-hub for what you're trying to do here. I'm sure he'd be happy to have your feedback as prospective users.

**Comment by asdofindia** [comment]
Looking at mozilla/mozfest-program#325, I think @phillipadsmith will be able to find us an easy way to implement this.

He already linked to:
- [redliner](https://github.com/benbalter/redliner) which I haven't explored much, but looks like an important part of what we'll need.

I've pretty much an idea of how it should look like at the end.

There'll be a "contribute to this page" button on top of every page. Clicking it shall give 3 options:
1. Edit anonymously
2. Sign in with github and edit
3. Report an issue (checkout a creative use of github issues [here](https://pantheon.io/docs/articles/sites/code/redirect-incoming-requests/))

These options will respectively
1. Immediately make the content editable with markdown, probably with a side by side preview.
2. Reach the same state as 1, after an authentication with github.
3. Lead to a pre-filled "new issue" page on github issues.

Options 1 & 2 will have a small box at the bottom to describe changes (commit message) and on clicking submit a pull request will be made and the link to it will be shown to the person.

(This will be made possible by a server that communicates with the github api, creates a commit, and then a pull request. Checkout [redliner](https://github.com/benbalter/redliner/), and wiki-hub [anonymous](https://github.com/wiki-hub/jekyll-anon) and [collaborative](https://github.com/wiki-hub/jekyll-collab) for ideas on how to make this server)

The changes will be visible on the portal only after the pull requests are merged. Optionally we can have the changes a user makes be stored in their localstorage so that if they edit the page again, they edit their modified version.

## 

I think that's what you have in mind @emmairwin. If anyone wants to start working on this spec, feel free to comment here. (I'm stuck at the authentication part and will be, for a few days at least). Of course, make sure @emmairwin comments agreeing with this spec before you start working.

**Comment by emmairwin** [comment]
@asdofindia this looks amazing.
Is there a way we can make this function as a plugin for Jekyll (if that's the right term) so we can share with others?   That's only a side note, obviously a nice to have that's all.

The ability to make small edits, is a huge first step for new curriculum contributors -  I know you raised this earlier as well 'why so much for small edits', but it's  providing a first step in contribution, the next steps will include writing new content as well.  But I get ahead of myself.

@asdofindia you rock!    If there are small pieces of coding you can break off, please do (and create sub issues for this )

**Comment by emmairwin** [comment]
also cc/ @soapdog

**Comment by soapdog** [comment]
@emmairwin thanks a lot for copying me here since this is a topic that really interest me. 

The approach mentioned here http://mdswanson.com/blog/2011/07/23/digging-around-the-github-api-take-2.html might help. This is an old post and the current API may be easier.

The main issue that will lead to friction is the authentication which I believe requires an OAuth workflow and might not be possible in a pure client-side application. I will try to do a Proof-of-Concept thing next week and see where we go from here.

**Comment by asdofindia** [comment]
Ah, just (re-?)discovered [gatekeeper](https://github.com/prose/gatekeeper). It was in [notes of getting down with github](http://etherpad.io/getting-down-with-github). It is server side code for the oauth. 

[This issue](https://github.com/prose/gatekeeper/issues/20) and [this blog post](http://blog.vjeux.com/2012/javascript/github-oauth-login-browser-side.html) explains how it works and there's a skeleton app [here](https://github.com/vjeux/GithubLogin).

@soapdog I hope these will be of use to you.

If we're storing the access token in localStorage though, we'll have to make sure there's a way for the user to not save the access token (for people who might be browsing from public computers).

**Comment by emmairwin** [comment]
@soapdog @asdofindia  Can't we just direct people to login at Github? And then check to see if there is an open session vrs authenticating ourselves?

**Comment by asdofindia** [comment]
@emmairwin I'm afraid that wouldn't be possible as it would have to violate the rules of [same origin policy](https://code.google.com/p/browsersec/wiki/Part2).

Just think of it like this. If I have an account on AshleyMadison, no other website should be able to know that I do have an account (or that I'm logged in). If that could happen, there'd be no privacy (or security).

**Comment by emmairwin** [comment]
Right. So lets drop Github pages from the equation.

**Comment by emmairwin** [comment]
Instead lets think about this workflow, and don't do any dev work because this is actually super important to planning for next year, and better to keep brainstorming here.  This is very rough, but thinking about how we can help people level up in github through contributing.  Vrs all or nothing.
1.  We have a UI that brings in Markdown curriculum from different repos
2.   The UI code(front end and back end) lives in one place, curriculum is 'included'.
3.   When curriculum is displayed, it has an 'improve this curriculum button
4.   When you click this button there are two choices - fork this repo , or edit in page.
5.  When you click 'Save' for edit, there are three more options
   -> Fork this Repo, Save Changes and Submit Pull Request
   -> Fork this Repo, Save Changes , I will submit my own pull request
   -> Fork this Repo, I will do the rest.

Something like this, of course fork only happens once this isn't the exact right workflow.  Ideas on this as something that teaches github, through a series of contributions?  How it is build would depend on platform, but lets have fun with it for now.  Assuing server side is available for authentication.

**Comment by asdofindia** [comment]
> Fork this repo, I'll do the rest 

This option after editing? Not required I guess.

Also since fork happens only once, step 4 is better written as 
- Edit here (easy)
- Edit on github (advanced)

Also, dropping or keeping github pages makes no change in this workflow. So we may as well keep it.

**Comment by emmairwin** [comment]
Edit here can also be broken down into 

1) saving to my Fork
2) create a pull request

**Comment by emmairwin** [comment]
Our curriculum direction in 2016 will depend on outcomes here:  https://github.com/mozilla/participation-org/issues/216
and will have more information after Orlando.

what will remain the same, and is a good challenge is to think about how we could design something that (at first) shields the contributor from complexities of github,  and then provides additional  contribution experiences that  peel back the layers and teach it.
